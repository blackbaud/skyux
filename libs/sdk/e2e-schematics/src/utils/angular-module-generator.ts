import { Tree } from '@nx/devkit';
import { Schema as ComponentSchema } from '@schematics/angular/component/schema';
import { Schema as ModuleSchema } from '@schematics/angular/module/schema';

import { wrapAngularDevkitSchematic } from 'nx/src/adapter/ngcli-adapter';

export async function angularModuleGenerator(tree: Tree, schema: ModuleSchema) {
  return wrapAngularDevkitSchematic('@schematics/angular', 'module')(
    tree,
    schema
  );
}

export async function angularComponentGenerator(
  tree: Tree,
  schema: ComponentSchema
) {
  return wrapAngularDevkitSchematic('@schematics/angular', 'component')(
    tree,
    schema
  );
}
