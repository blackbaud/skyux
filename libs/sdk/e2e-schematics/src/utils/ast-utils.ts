import { getSourceNodes } from '@angular/cdk/schematics';
import { Tree } from '@nx/devkit';
import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

export type DecoratedClass = {
  classDeclaration: ts.ClassDeclaration;
  properties: { [key: string]: ts.Expression };
  propertiesObjectLiteral?: ts.ObjectLiteralExpression;
};

/**
 * Apply a set of transformations to typescript source files.
 */
export function applyTransformers(
  sourceFiles: ts.SourceFile[],
  transformers: ts.TransformerFactory<ts.SourceFile>[]
): ts.SourceFile[] {
  const transformation = ts.transform(sourceFiles, transformers);
  return transformation.transformed;
}

/**
 * Apply a set of transformations to a typescript source file in a @nx/devkit tree.
 */
export function applyTransformersToPath(
  tree: Tree,
  sourceFile: string,
  transformers: ts.TransformerFactory<ts.SourceFile>[]
): void {
  const sourceFiles = readSourceFile(tree, sourceFile);
  const [result] = applyTransformers([sourceFiles], transformers);
  writeSourceFile(tree, sourceFile, result);
}

// From https://github.com/nrwl/nx/blob/master/packages/angular/src/generators/utils/insert-ngmodule-import.ts
export function findImport(
  sourceFile: ts.SourceFile,
  importPath: string
): ts.ImportDeclaration | undefined {
  const importStatements = sourceFile.statements.filter(ts.isImportDeclaration);
  return importStatements.find(
    (statement) =>
      statement.moduleSpecifier
        .getText(sourceFile)
        .replace(/['"`]/g, '')
        .trim() === importPath.replace(/\.ts$/, '')
  );
}

// From https://github.com/nrwl/nx/blob/master/packages/angular/src/generators/utils/insert-ngmodule-import.ts
export function getNamedImport(
  coreImport: ts.ImportDeclaration,
  importName: string
): ts.ImportSpecifier | undefined {
  if (!coreImport.importClause) {
    throw new Error(`Could not find an import.`);
  }

  if (
    !(
      !!coreImport.importClause.namedBindings &&
      ts.isNamedImports(coreImport.importClause.namedBindings)
    )
  ) {
    throw new Error(
      `The import from ${
        (coreImport.moduleSpecifier as ts.StringLiteral).text
      } does not have named imports.`
    );
  }

  return coreImport.importClause.namedBindings.elements.find((namedImport) =>
    namedImport.propertyName
      ? ts.isIdentifier(namedImport.propertyName) &&
        namedImport.propertyName.escapedText === importName
      : ts.isIdentifier(namedImport.name) &&
        namedImport.name.escapedText === importName
  );
}

// From https://github.com/nrwl/nx/blob/master/packages/angular/src/generators/utils/insert-ngmodule-import.ts
function findDecoratedClass(
  sourceFile: ts.SourceFile,
  ngModuleName: ts.__String
): DecoratedClass | undefined {
  const classDeclarations = sourceFile.statements.filter(ts.isClassDeclaration);
  const classDeclaration = classDeclarations.find((declaration) => {
    const decorators = ts.getDecorators(declaration);
    return decorators
      ? decorators.some(
          (decorator) =>
            ts.isCallExpression(decorator.expression) &&
            ts.isIdentifier(decorator.expression.expression) &&
            decorator.expression.expression.escapedText === ngModuleName
        )
      : undefined;
  });
  if (classDeclaration) {
    const decorators = ts.getDecorators(classDeclaration);

    const decorator = decorators?.find(
      (decorator) =>
        ts.isCallExpression(decorator.expression) &&
        ts.isIdentifier(decorator.expression.expression) &&
        decorator.expression.expression.escapedText === ngModuleName
    );

    const properties: { [key: string]: ts.Expression } = {};

    const callExpression = decorator?.expression as ts.CallExpression;

    let propertiesObjectLiteral: ts.ObjectLiteralExpression | undefined;

    if (callExpression.arguments.length > 0) {
      if (!ts.isObjectLiteralExpression(callExpression.arguments[0])) {
        throw new Error(
          `The ${ngModuleName} options for ${classDeclaration.name?.escapedText} are not an object literal`
        );
      }

      propertiesObjectLiteral = callExpression
        .arguments[0] as ts.ObjectLiteralExpression;
      propertiesObjectLiteral.properties.forEach((property) => {
        if (
          ts.isPropertyAssignment(property) &&
          ts.isIdentifier(property.name)
        ) {
          properties[property.name.text] = property.initializer;
        }
      });
    }
    return { classDeclaration, propertiesObjectLiteral, properties };
  }
  return undefined;
}

export function findNgModuleClass(
  sourceFile: ts.SourceFile
): DecoratedClass | undefined {
  const coreImport = findImport(sourceFile, '@angular/core');
  if (!coreImport) {
    throw new Error('Could not find @angular/core import.');
  }
  const ngModuleClass = getNamedImport(coreImport, 'NgModule');
  if (!ngModuleClass) {
    throw new Error('Could not find NgModule import.');
  }
  return findDecoratedClass(sourceFile, ngModuleClass.name.escapedText);
}

export function findComponentClass(
  sourceFile: ts.SourceFile
): DecoratedClass | undefined {
  const coreImport = findImport(sourceFile, '@angular/core');
  if (!coreImport) {
    return undefined;
  }
  const componentClass = getNamedImport(coreImport, 'Component');
  if (!componentClass) {
    return undefined;
  }
  return findDecoratedClass(sourceFile, componentClass.name.escapedText);
}

/**
 * Transform a parsed typescript source file into a string.
 */
export function getSourceAsString(sourceFile: ts.Node): string {
  return (
    ts
      .createPrinter({
        newLine: ts.NewLineKind.LineFeed,
      })
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .printNode(ts.EmitHint.Unspecified, sourceFile, undefined as any)
  );
}

/**
 * Read the value of a string literal from a typescript source file.
 */
export function getStringLiteral(
  sourceFile: ts.SourceFile,
  name: string
): string {
  const nodes = getSourceNodes(sourceFile);
  const identifierNode = nodes.find(
    (node) => ts.isIdentifier(node) && (node as ts.Identifier).text === name
  );
  if (
    identifierNode &&
    identifierNode.parent &&
    [
      ts.SyntaxKind.VariableDeclaration,
      ts.SyntaxKind.PropertyAssignment,
    ].includes(identifierNode.parent.kind)
  ) {
    const value = identifierNode.parent.getLastToken();
    if (value && ts.isStringLiteral(value)) {
      return (value as ts.StringLiteral).text;
    }
  }
  throw new Error(`Unable to find ${name}`);
}

/**
 * Create a transformer to add an export statement to a typescript source file.
 */
export function getInsertExportTransformer(
  insertPath: string,
  afterPath: string
): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const visitor: (
        node: ts.Node
      ) => ts.Node | ts.NodeArray<ts.ExportDeclaration> = (
        rootNode: ts.Node
      ) => {
        const node = ts.visitEachChild(rootNode, visitor as any, context);
        if (ts.isExportDeclaration(node)) {
          const exportPath = (
            (node as ts.ExportDeclaration).moduleSpecifier as ts.StringLiteral
          ).text;
          if (exportPath === afterPath) {
            return context.factory.createNodeArray([
              node,
              context.factory.createExportDeclaration(
                undefined,
                undefined,
                false,
                undefined,
                context.factory.createStringLiteral(insertPath, true)
              ),
            ]);
          }
        }
        return node;
      };
      return ts.visitNode(sourceFile, visitor as any);
    };
  };
}

export function getInsertImportTransformer(
  symbol: string,
  path: string
): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const importStatement = findImport(sourceFile, path);
      if (importStatement && getNamedImport(importStatement, symbol)) {
        // Already imported
        return sourceFile;
      }
      // Add import statement to the top of the file. Prettier will sort this out.
      return context.factory.updateSourceFile(sourceFile, [
        context.factory.createImportDeclaration(
          [],
          context.factory.createImportClause(
            false,
            undefined,
            context.factory.createNamedImports([
              context.factory.createImportSpecifier(
                false,
                undefined,
                context.factory.createIdentifier(symbol)
              ),
            ])
          ),
          context.factory.createStringLiteral(path, true)
        ),
        ...sourceFile.statements,
      ]);
    };
  };
}

/**
 * Create a transformer to add an identifier to an array in a typescript source file.
 */
export function getInsertIdentifierToArrayTransformer(
  propertyName: string,
  identifier: string
): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceNode: ts.SourceFile) => {
      const visitor: (node: ts.Node) => ts.Node | ts.PropertyAssignment = (
        rootNode: ts.Node
      ) => {
        const node = ts.visitEachChild(rootNode, visitor, context);
        if (
          ts.isPropertyAssignment(node) &&
          (node as ts.PropertyAssignment).name.getText() === propertyName &&
          ts.isArrayLiteralExpression(
            (node as ts.PropertyAssignment).initializer
          )
        ) {
          return context.factory.createPropertyAssignment(
            propertyName,
            context.factory.createArrayLiteralExpression([
              ...(
                (node as ts.PropertyAssignment)
                  .initializer as ts.ArrayLiteralExpression
              ).elements,
              context.factory.createIdentifier(identifier),
            ])
          );
        }
        return node;
      };
      return ts.visitNode(sourceNode, visitor);
    };
  };
}

/**
 * Create a transformer to add an identifier to an NgModule decorator property in a typescript source file.
 */
function getTransformerToAddSymbolToDecoratedClassMetadata(
  importPath: string,
  importName: string,
  metadataField: string,
  expression: string
): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const coreImport = findImport(sourceFile, importPath);
      if (!coreImport) {
        return sourceFile;
      }
      const ngModuleClass = getNamedImport(coreImport, importName);
      if (!ngModuleClass) {
        return sourceFile;
      }
      const decoratedClass = findDecoratedClass(
        sourceFile,
        ngModuleClass.name.escapedText
      );
      if (!decoratedClass) {
        return sourceFile;
      }
      if (metadataField in decoratedClass.properties) {
        if (
          !ts.isArrayLiteralExpression(decoratedClass.properties[metadataField])
        ) {
          // Unexpected metadata field type
          return sourceFile;
        }
        const existingExpressions = (
          decoratedClass.properties[metadataField] as ts.ArrayLiteralExpression
        ).elements.map((element) =>
          getSourceAsString(element).trim().replace(/;$/, '')
        );
        if (existingExpressions.includes(expression)) {
          // Already added
          return sourceFile;
        }
        const newExpression = context.factory.createIdentifier(expression);
        const appendValueVisitor: (
          node: ts.Node
        ) => ts.Node | ts.ArrayLiteralExpression = (rootNode: ts.Node) => {
          if (
            ts.isArrayLiteralExpression(rootNode) &&
            rootNode === decoratedClass.properties[metadataField]
          ) {
            return context.factory.updateArrayLiteralExpression(rootNode, [
              ...rootNode.elements,
              newExpression,
            ]);
          }
          return ts.visitEachChild(rootNode, appendValueVisitor, context);
        };
        return ts.visitNode(sourceFile, appendValueVisitor);
      } else {
        const newExpression = context.factory.createPropertyAssignment(
          metadataField,
          context.factory.createArrayLiteralExpression([
            context.factory.createIdentifier(expression),
          ])
        );
        const newPropertyVisitor: (
          node: ts.Node
        ) => ts.Node | ts.ObjectLiteralExpression = (rootNode: ts.Node) => {
          if (
            ts.isObjectLiteralExpression(rootNode) &&
            rootNode === decoratedClass.propertiesObjectLiteral
          ) {
            return context.factory.updateObjectLiteralExpression(rootNode, [
              ...rootNode.properties,
              newExpression,
            ]);
          }
          return ts.visitEachChild(rootNode, newPropertyVisitor, context);
        };
        return ts.visitNode(sourceFile, newPropertyVisitor);
      }
    };
  };
}

/**
 * Create a transformer to add a component to an NgModule export.
 */
export function getTransformerToAddExportToNgModule(
  identifier: string
): ts.TransformerFactory<ts.SourceFile> {
  return getTransformerToAddSymbolToDecoratedClassMetadata(
    '@angular/core',
    'NgModule',
    'exports',
    identifier
  );
}

/**
 * Create a transformer to add a string literal value to an object literal in a typescript source file.
 */
export function getInsertStringPropertyTransformer(
  beforeProperty: string,
  name: string,
  value: string
): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const visitor: (node: ts.Node) => ts.Node | ts.Identifier = (
        rootNode: ts.Node
      ) => {
        const node = ts.visitEachChild(rootNode, visitor, context);
        if (ts.isObjectLiteralExpression(node)) {
          const beforeChild = (
            node as ts.ObjectLiteralExpression
          ).properties.findIndex(
            (property) =>
              ts.isPropertyAssignment(property) &&
              ts.isIdentifier(property.name) &&
              property.name.text === beforeProperty
          );
          if (beforeChild > -1) {
            const properties = Array.from(
              (node as ts.ObjectLiteralExpression).properties.values()
            );
            properties.splice(
              beforeChild,
              0,
              context.factory.createPropertyAssignment(
                name,
                context.factory.createStringLiteral(value)
              )
            );
            return context.factory.updateObjectLiteralExpression(
              node as ts.ObjectLiteralExpression,
              properties
            );
          }
        }
        return node;
      };
      return ts.visitNode(sourceFile, visitor);
    };
  };
}

/**
 * Create a transformer to update the value of string literals in a typescript source file.
 */
export function getStringLiteralsSetterTransformer(strings: {
  [_: string]: string;
}): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const visitor: (node: ts.Node) => ts.Node | ts.Identifier = (
        rootNode: ts.Node
      ) => {
        const node = ts.visitEachChild(rootNode, visitor, context);
        if (
          node.parent &&
          ts.isStringLiteral(node) &&
          ts.isIdentifier(node.parent.getChildAt(0)) &&
          (node.parent.getChildAt(0) as ts.Identifier).text
        ) {
          const identifierName = (
            node.parent.getChildAt(0) as ts.Identifier
          ).getText(sourceFile);
          if (identifierName in strings) {
            return context.factory.createStringLiteral(strings[identifierName]);
          }
        }
        return node;
      };
      return ts.visitNode(sourceFile, visitor);
    };
  };
}

/**
 * Create a transformer to rename variables in a typescript source file.
 */
export function getRenameVariablesTransformer(renameMap: {
  [_: string]: string;
}): ts.TransformerFactory<ts.SourceFile> {
  return (context) => {
    return (sourceFile: ts.SourceFile) => {
      const visitor: (node: ts.Node) => ts.Node | ts.Identifier = (
        rootNode: ts.Node
      ) => {
        const node = ts.visitEachChild(rootNode, visitor, context);
        if (ts.isIdentifier(node) && (node as ts.Identifier).text) {
          const identifierName = (node as ts.Identifier).getText(sourceFile);
          if (identifierName in renameMap) {
            return context.factory.createIdentifier(renameMap[identifierName]);
          }
        }
        return node;
      };
      return ts.visitNode(sourceFile, visitor);
    };
  };
}

/**
 * Read a typescript source file from a @nx/devkit tree.
 */
export function readSourceFile(
  tree: Tree,
  path: string,
  preprocess: (source: string) => string = (source) => source
): ts.SourceFile {
  const sourceText = tree.read(path, 'utf-8') ?? '';
  return ts.createSourceFile(
    path,
    preprocess(sourceText),
    ts.ScriptTarget.Latest,
    true
  );
}

/**
 * Write a typescript source file to a @nx/devkit tree.
 */
export function writeSourceFile(
  tree: Tree,
  path: string,
  sourceFile: ts.SourceFile
): void {
  tree.write(path, getSourceAsString(sourceFile));
}
