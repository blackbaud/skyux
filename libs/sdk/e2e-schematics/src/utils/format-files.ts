import { Tree, formatFiles as nxFormatFiles } from '@nx/devkit';

export async function formatFiles(
  tree: Tree,
  options?: { skipFormat?: boolean },
): Promise<void> {
  if (!options?.skipFormat) {
    await nxFormatFiles(tree);
  }
}
