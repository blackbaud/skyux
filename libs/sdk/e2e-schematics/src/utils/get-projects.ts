import { ProjectConfiguration, Tree, getProjects } from '@nrwl/devkit';

function filterProjects(
  tree: Tree,
  filter: (projectConfiguration: ProjectConfiguration) => boolean,
  name?: string
): Map<string, ProjectConfiguration> {
  const projectConfigurations = getProjects(tree);
  const projects = new Map<string, ProjectConfiguration>();
  if (name) {
    const projectNames = name
      .split(',')
      .map((projectName) => projectName.trim())
      .filter(
        (projectName) => projectName && projectConfigurations.has(projectName)
      );
    projectNames.forEach((projectName) => {
      const projectConfiguration = projectConfigurations.get(projectName);
      if (projectConfiguration && filter(projectConfiguration)) {
        projects.set(projectName, projectConfiguration);
      }
    });
  } else {
    projectConfigurations.forEach((projectConfiguration, projectName) => {
      if (filter(projectConfiguration)) {
        projects.set(projectName, projectConfiguration);
      }
    });
  }
  return projects;
}

function projectHasTargetWithExecutor(
  projectConfiguration: ProjectConfiguration,
  executor: string
): boolean {
  return (
    !!projectConfiguration.targets &&
    Object.values(projectConfiguration.targets).some(
      (target) => target.executor === executor
    )
  );
}

export function getE2eProjects(
  tree: Tree,
  name?: string
): Map<string, ProjectConfiguration> {
  return filterProjects(
    tree,
    (projectConfiguration) =>
      projectHasTargetWithExecutor(
        projectConfiguration,
        '@nrwl/cypress:cypress'
      ),
    name
  );
}

export function getStorybookProjects(
  tree: Tree,
  name?: string
): Map<string, ProjectConfiguration> {
  return filterProjects(
    tree,
    (projectConfiguration) =>
      projectHasTargetWithExecutor(
        projectConfiguration,
        '@storybook/angular:build-storybook'
      ),
    name
  );
}

export function getStorybookProject(
  tree: Tree,
  options: Partial<{ project: string }>
) {
  if (!options.project) {
    throw new Error(`Project name not specified`);
  }

  const projects = getProjects(tree);
  if (!projects.has(options.project)) {
    throw new Error(`Unable to find project ${options.project}`);
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  let projectConfig = projects.get(options.project)!;

  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  if (!('storybook' in projectConfig.targets!)) {
    options.project = `${options.project}-storybook`;
    if (!projects.has(options.project)) {
      throw new Error(`Unable to find project ${options.project}`);
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    projectConfig = projects.get(options.project)!;

    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    if (!('storybook' in projectConfig.targets!)) {
      throw new Error(`Storybook is not configured for ${options.project}`);
    }
  }
  return projectConfig;
}

export function getProjectTypeBase(projectConfig: ProjectConfiguration) {
  return `${projectConfig.projectType == 'library' ? 'lib' : 'app'}`;
}
