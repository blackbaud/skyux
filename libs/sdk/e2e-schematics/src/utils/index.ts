export * from './angular-module-generator';
export * from './ast-utils';
export * from './find-module';
export * from './get-projects';
export * from './update-json';
export * from './utils';
