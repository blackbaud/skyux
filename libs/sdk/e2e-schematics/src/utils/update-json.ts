import { Tree, parseJson, serializeJson } from '@nrwl/devkit';

export function readJsonFile(tree: Tree, path: string) {
  const asString = tree.read(path, 'utf-8');

  if (asString) {
    return parseJson(asString);
  } else {
    return {};
  }
}

function writeJsonFile(tree: Tree, path: string, data: unknown) {
  const asString = serializeJson(data as never);
  tree.write(path, asString);
}

export function updateJson<T>(tree: Tree, path: string, callback: (a: T) => T) {
  const json = readJsonFile(tree, path);
  callback(json);
  writeJsonFile(tree, path, json);
}
