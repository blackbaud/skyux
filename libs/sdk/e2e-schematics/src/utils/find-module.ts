import { Tree, normalizePath, visitNotIgnoredFiles } from '@nx/devkit';
import * as ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

import { relative } from 'path';

import {
  DecoratedClass,
  findComponentClass,
  findImport,
  findNgModuleClass,
  getNamedImport,
  readSourceFile,
} from './ast-utils';
import { dirname } from './utils';

export function findModulePaths(
  tree: Tree,
  path: string,
  predicate: (path: string) => boolean,
  endsWith = '.module.ts'
): string[] {
  const modulePaths: string[] = [];
  visitNotIgnoredFiles(tree, path, (filePath) => {
    if (filePath.endsWith(endsWith)) {
      if (predicate(filePath)) {
        modulePaths.push(normalizePath(filePath));
      }
    }
  });
  return modulePaths;
}

export function findDeclaringModule(
  tree: Tree,
  path: string,
  componentPath: string
): { filepath: string; module: DecoratedClass } | null {
  let result: { filepath: string; module: DecoratedClass } | null = null;
  const componentSource = readSourceFile(tree, componentPath);
  const componentClass = findComponentClass(componentSource);
  if (componentClass) {
    findModulePaths(
      tree,
      path,
      (filepath) => {
        filepath = normalizePath(filepath);
        const filepathDirectory = filepath.substring(
          0,
          filepath.lastIndexOf('/')
        );
        let relativePath = normalizePath(
          relative(filepathDirectory, componentPath)
        );
        if (!relativePath.startsWith('.')) {
          relativePath = `./${relativePath}`;
        }
        try {
          const sourceFile = readSourceFile(tree, filepath);
          const componentFileImport = findImport(sourceFile, relativePath);
          if (!componentFileImport) {
            return false;
          }
          const componentNamedImport = getNamedImport(
            componentFileImport,
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            componentClass.classDeclaration!.name!.text
          );
          const module = findNgModuleClass(sourceFile);
          if (module) {
            if (
              module.properties['declarations'] &&
              ts.isArrayLiteralExpression(module.properties['declarations'])
            ) {
              const declarations = module.properties['declarations'].elements;
              for (const declaration of declarations) {
                if (
                  ts.isIdentifier(declaration) &&
                  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                  declaration.text === componentNamedImport!.name.text
                ) {
                  result = { filepath, module };
                  return true;
                }
              }
            }
          }
          // eslint-disable-next-line no-empty
        } catch (e) {}
        return false;
      },
      '.module.ts'
    );
  }
  return result;
}

export function isRoutingModule(
  module: DecoratedClass,
  sourceFile: ts.SourceFile
): boolean {
  // Is one of the imports a static call to a method on RouterModule?
  if (
    module.properties['imports'] &&
    ts.isArrayLiteralExpression(module.properties['imports'])
  ) {
    if (
      (module.properties['imports'] as ts.ArrayLiteralExpression).elements.some(
        (im) => {
          if (
            ts.isCallExpression(im) &&
            ts.isPropertyAccessExpression(im.getChildren(sourceFile)[0])
          ) {
            const propertyAccessExpression = im.getChildren(
              sourceFile
            )[0] as ts.PropertyAccessExpression;
            if (
              ts.isIdentifier(
                (propertyAccessExpression as ts.PropertyAccessExpression)
                  .expression
              )
            ) {
              return (
                (propertyAccessExpression.expression as ts.Identifier).text ===
                'RouterModule'
              );
            }
          }
          return false;
        }
      )
    ) {
      return true;
    }
  }
  return false;
}

export function findClosestModule(
  modulePaths: string[],
  projectDirectory: string,
  closestToDirectory: string
): string | undefined {
  projectDirectory = normalizePath(projectDirectory);
  closestToDirectory = normalizePath(closestToDirectory);
  const modulePathsNormalized = modulePaths.map((modulePath) =>
    normalizePath(modulePath)
  );
  modulePathsNormalized.sort((a, b) => {
    const aDir = dirname(a);
    const bDir = dirname(b);
    if (aDir.length === bDir.length) {
      return a > b ? 1 : -1;
    }
    return aDir.length < bDir.length ? 1 : -1;
  });
  let directory = `${projectDirectory}/${closestToDirectory}`;
  while (
    directory &&
    directory !== projectDirectory &&
    !modulePathsNormalized.some((m) => dirname(m) === directory)
  ) {
    directory = dirname(directory);
  }
  if (!directory || directory === projectDirectory) {
    return undefined;
  }
  // Return the basename of the module.
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return modulePathsNormalized
    .find((m) => dirname(m) === directory)!
    .split('/')
    .pop()!
    .replace(/\.module\.ts$/, '');
}
