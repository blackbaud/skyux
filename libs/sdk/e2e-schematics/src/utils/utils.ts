import { decamelize } from '@nx/workspace/src/utils/strings';

export function dirname(path: string): string {
  const end = path.lastIndexOf('/');
  return path.substring(0, end);
}

export function capitalizeWords(string: string) {
  return decamelize(string)
    .replace(/[-_]/g, ' ')
    .replace(/\b[a-z]/g, (char) => char.toUpperCase());
}
