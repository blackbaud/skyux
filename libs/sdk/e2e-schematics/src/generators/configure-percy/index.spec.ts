import { cypressProjectGenerator } from '@nrwl/cypress';

import { createTreeWithEmptyWorkspace } from 'nx/src/generators/testing-utils/create-tree-with-empty-workspace';

import { updateJson } from '../../utils/update-json';

import configurePercy from './index';

describe('configure-percy', () => {
  it('should import percy', async () => {
    const tree = createTreeWithEmptyWorkspace(1);
    await cypressProjectGenerator(tree, {
      name: `cypress`,
      baseUrl: 'https://example.com',
    });
    expect(tree.exists('apps/cypress/src/support/index.ts')).toBeTruthy();
    expect(
      tree.read('apps/cypress/src/support/index.ts').toString()
    ).not.toContain('percy');
    await configurePercy(tree, { name: 'cypress' });
    expect(tree.read('apps/cypress/src/support/index.ts').toString()).toContain(
      'percy'
    );
  });

  it('should not make changes without cypress.json', async () => {
    const tree = createTreeWithEmptyWorkspace(1);
    await cypressProjectGenerator(tree, {
      name: `cypress`,
      baseUrl: 'https://example.com',
    });
    tree.delete(`apps/cypress/cypress.json`);
    const beforeLength = tree.listChanges().length;
    await configurePercy(tree, { name: 'cypress' });
    expect(tree.listChanges().length).toEqual(beforeLength);
  });

  it('should handle missing supportFile config', async () => {
    const tree = createTreeWithEmptyWorkspace(1);
    await cypressProjectGenerator(tree, {
      name: `cypress`,
      baseUrl: 'https://example.com',
    });
    updateJson<any>(tree, `apps/cypress/cypress.json`, (config) => {
      delete config.supportFile;
      return config;
    });
    await configurePercy(tree, { name: 'cypress' });
    const config = JSON.parse(
      tree.read(`apps/cypress/cypress.json`).toString()
    );
    expect(config.video).toBeFalsy();
  });

  it('should handle missing supportFile', async () => {
    const tree = createTreeWithEmptyWorkspace(1);
    await cypressProjectGenerator(tree, {
      name: `cypress`,
      baseUrl: 'https://example.com',
    });
    tree.delete(`apps/cypress/src/support/index.ts`);
    await configurePercy(tree, { name: 'cypress' });
    expect(tree.exists(`apps/cypress/src/support/index.ts`)).toBeTruthy();
  });
});
