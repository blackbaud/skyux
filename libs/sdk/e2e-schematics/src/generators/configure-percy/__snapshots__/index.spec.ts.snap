// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configure-percy should generate cypress.config.ts 1`] = `
"import { nxE2EPreset } from '@nx/cypress/plugins/cypress-preset';
import { defineConfig } from 'cypress';

export default defineConfig({
  e2e: {
    ...nxE2EPreset(__dirname),
    video: false,
    defaultCommandTimeout: 60000,
  },
});
"
`;
